homework/shared_ptr/my_shared_ptr.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+#include <atomic>
+#include <iostream>
+
+struct Counter
+{
+    std::atomic<int> count_ = 0;
+};
+
+template <typename T>
+class my_shared_ptr
+{
+
+public:
+    explicit my_shared_ptr(T* ptr) : ptr_(ptr)
+    {
+        if (counter_ptr != nullptr)
+        {
+            counter_ptr->count_++;
+        }
+        else
+        {
+            counter_ptr = new Counter();
+            counter_ptr->count_++;
+        }
+    };
+
+    my_shared_ptr() : counter_ptr(new Counter()) { ptr_ = nullptr; };
+
+    my_shared_ptr(const my_shared_ptr<T>& other_shared_ptr)
+    {
+        this->ptr_ = other_shared_ptr.ptr_;
+        this->counter_ptr = other_shared_ptr.counter_ptr;
+
+        counter_ptr->count_++;
+    };
+
+    my_shared_ptr(my_shared_ptr<T>&& ptr_moved)
+    {
+        ptr_ = ptr_moved.get();
+        if (counter_ptr != nullptr)
+        {
+            counter_ptr->count_ = ptr_moved.use_count();
+        }
+        else
+        {
+            counter_ptr = new Counter();
+            counter_ptr->count_ = ptr_moved.use_count();
+        }
+        ptr_moved.reset(nullptr);
+    }
+
+    T& operator=(const my_shared_ptr<T>& other_shared_ptr)
+    {
+        std::cout << "operator move assigment" << "\n";
+        delete ptr_;
+        delete counter_ptr;
+
+        this->ptr_ = other_shared_ptr.ptr_;
+        this->counter_ptr = other_shared_ptr.counter_ptr;
+        if (other_shared_ptr)
+        {
+            this->counter_ptr->count_++;
+        }
+    };
+
+    T& operator=(const my_shared_ptr<T>&& some_ptr)
```

Comment:
```suggestion
    T& operator=(my_shared_ptr<T>&& some_ptr)
```
move assignment operator takes non-const rvalue ref